### ISSUE TITLE
CS8762 false positive on expression return

### ISSUE THREAD

**Issue by saucecontrol** [dotnet/roslyn#42855]
**Version Used**: 16.6 Preview 2

**Steps to Reproduce**:

The following method just started giving a false positive CS8762 (Parameter 'value' may not have a null value when exiting with 'true') on VS 16.6p2

```C#
public static bool TryGetMetadataByName(this IWICMetadataQueryReader meta, string name, [NotNullWhen(true)] out PropVariant? value)
{
    value = null;

    int hr = ProxyFunctions.GetMetadataByName(meta, name, IntPtr.Zero);
    if (hr >= 0)
    {
        value = new PropVariant();

        var pvMarshal = PropVariant.Marshaler.GetInstance(null);
        var pvNative = pvMarshal.MarshalManagedToNative(value);
        hr = ProxyFunctions.GetMetadataByName(meta, name, pvNative);
        pvMarshal.MarshalNativeToManaged(pvNative);
        pvMarshal.CleanUpNativeData(pvNative);
    }

    return hr >= 0;  // <-- warning CS8762 here
}
```

In this case, the return can't be true unless the expression evaluated to true in the `if`, and `value` would have been set to non-null

**Comment by gafter** [comment]
@jcouv I think this is by design but I'm not sure what advice to offer the programmer.

**Comment by jcouv** [comment]
I would add an assertion

**Comment by scottdorman** [comment]
This seems like it may be related (if not, I'll file it as a new issue).

Given the following code:

```
internal static bool TryCastExact<T>(this IApiResult source, [NotNullWhen(true)] out T value) where T : class
{
    var success = false;
    value = default;

    if (source is T parsedValue)
    {
	value = parsedValue;
	success = true;
    }

    return success;
}
```

The compiler issues a warning on `return success;`

However, if that code is changed to

```
internal static bool TryCastExact<T>(this IApiResult source, [NotNullWhen(true)] out T value) where T : class
{
    value = default;

    if (source is T parsedValue)
    {
	value = parsedValue;
	return true;
    }

    return false;
}
```

No warning is generated. The two code snippets are, in effect, the same thing. The only difference is that the one which causes the warning has a single exit point while the one that doesn't cause the warning has multiple exit points. The analysis to generate the warning should be able to recognize both of these patterns and come up with the same decision.

Interestingly enough, this code also doesn't trigger a warning:

```
internal static bool TryCastExact<T>(this IApiResult source, [MaybeNullWhen(false)] out T? value) where T : class
{
    var success = false;
    value = default;

    if (source is T parsedValue)
    {
	value = parsedValue;
	success = true;
    }

    return success;
}
```

Although that doesn't seem to match the expectation. If the method returns false, the `value` will be `default` and since `T` is constrained to be a class (and declared as `T?`), it's going to be `null`, so saying `MayBeNullWhen(false)` isn't correct...it's always going to be `null` when false.

**Comment by saucecontrol** [comment]
I've just tried this again after noticing some other `NotNullWhen` cases were fixed. This case is no longer warning on VS 16.9p2 and SDK 5.0.200-preview.20601.7.
